Civi/Gpapi/ContractHelper/Adyen.php

```diff
@@ -2,150 +2,312 @@
+    ];
+  }
+
+  protected function getPaymentLabel() {
+    $account_number = $this->paymentToken['masked_account_number'];
+    $masked_digits = preg_replace('/\d/', '*', substr($account_number, 0, -4));
+    $last_digits = substr($account_number, -4);
+    $masked_account_number = $masked_digits . $last_digits;
+
+    $expiry_date = new DateTimeImmutable($this->paymentToken['expiry_date']);
+    $fmt_exp_date = $expiry_date->format('m/Y');
+
+    return "$masked_account_number ($fmt_exp_date)";
+  }
+
+  protected function getPspName() {
+    return self::PSP_NAME;
+  }
+
+  protected function loadAdditionalPaymentData() {
+    $this->paymentProcessor = Api4\PaymentProcessor::get(FALSE)
+      ->addWhere('id', '=', $this->recurringContribution['payment_processor_id'])
+      ->addSelect('*')
+      ->execute()
+      ->first();
+
+    $this->paymentToken = Api4\PaymentToken::get(FALSE)
+      ->addWhere('id', '=', $this->recurringContribution['payment_token_id'])
+      ->addSelect('*')
+      ->execute()
+      ->first();
+  }
+
+  private static function calculateNextDebitDate(array $params) {
+    $next_debit_date = strtotime('+1 month');
+    $cycle_day = (int) CRM_Utils_Array::value('cycle_day', $params, 0);
+
+    if (empty($cycle_day)) {
+      $possible_cycle_days = \CRM_Contract_PaymentAdapter_Adyen::cycleDays();
+      $cycle_day = (int) date('d', $next_debit_date);
+
+      while (!in_array($cycle_day, $possible_cycle_days)) {
+        $next_debit_date = strtotime("+ 1 day", $next_debit_date);
+        $cycle_day = (int) date('d', $next_debit_date);
+      }
+    } else {
+      while ((int) date('d', $next_debit_date) !== $cycle_day) {
+        $next_debit_date = strtotime("+ 1 day", $next_debit_date);
+      }
     }
-    else {
-      $creditorLookup['currency'] = $params['currency'];
+
+    return $next_debit_date;
+  }
+
+  private static function getAccountNumber(array $additional_psp_data) {
+    $card_summary = CRM_Utils_Array::value('cardSummary', $additional_psp_data);
+    $pm_variant = CRM_Utils_Array::value('paymentMethodVariant', $additional_psp_data);
+
+    if (empty($card_summary) || empty($pm_variant)) return NULL;
+
+    return ucfirst($pm_variant) . ": $card_summary";
+  }
+
+  private static function getCardHolderName(array $additional_psp_data) {
+    if (empty($additional_psp_data['cardHolderName'])) return [NULL, NULL];
+
+    $cardHolderName = $additional_psp_data['cardHolderName'];
+
+    if ($cardHolderName === 'Checkout Shopper PlaceHolder') return [NULL, NULL];
+
+    $cardHolderName = explode(' ', $cardHolderName);
+
+    if (count($cardHolderName) < 2) return [$cardHolderName[0], NULL];
+
+    $firstName = reset($cardHolderName);
+    $lastName = end($cardHolderName);
+
+    return [
+      empty($firstName) ? NULL : $firstName,
+      empty($lastName) ? NULL : $lastName,
+    ];
+  }
+
+  private static function getExpiryDate(array $additional_psp_data) {
```

Comment:
We should use the last day of the month for expiry date - this is how the adyen extension calculates it, in case you want to copy:
```
        if (preg_match('@^(\d\d)/(\d{4})$@', $this->eventData['additionalData']['expiryDate'], $matches)) {
          // cards expire at the end of this month.
          $expectedExpiry = date('Y-m-d H:i:s', strtotime("$matches[2]-$matches[1]-01 + 1 month - 1 minute"));
          if ($expectedExpiry !== $result['payment_token_id.expiry_date']) {
            $updates['expiry_date'] = $expectedExpiry;
          }
        }
```